# STAT 540: Seminar 7
# RNA-seq Analysis: Getting Read Counts
# Date: Wed Feb 26
# By: Jessica Pilsworth
========================================================

Load libraries
```{r}
library(ShortRead)
library(Rsamtools)
library(easyRNASeq)
```

Load data
```{r}
bamDat <- readAligned("drosophilaMelanogasterSubset.bam", 
                      type = "BAM")
str(bamDat)
```

To index a bam file
```{r}
indexFile <- indexBam("drosophilaMelanogasterSubset.bam")
```

To filter reads with at most 2N's
```{r}
nFilt <- nFilter(2)
```

To filter reads which have been aligned to the reference genome
```{r}
chrFilt <- chromosomeFilter(regex = "chr")
```

To create a new filter which checks that both conditions are satisfied i.e. a read has 2 or fewer N's and is aligned to the reference genome
```{r}
filt <- compose(nFilt, chrFilt)
```

Apply the filter to extract the relevant subset of data
```{r}
bamDatFiltered <- bamDat[filt(bamDat)]
```

To examine the filter bam file
```{r}
str(bamDatFiltered)
```

To look at which chromosomes are present and read IDs
```{r}
levels(chromosome(bamDatFiltered))
id(bamDatFiltered)[1:10]
```

To view the DNA sequence for the read
```{r}
sread(bamDatFiltered)[1:10]
```

To look at the base qualities
```{r}
quality(bamDatFiltered)[1:10]
```

To find out the starting (left most) position where the read was aligned on the chromosome
```{r}
position(bamDatFiltered)[1:10]
```

To see which strand (forward/+ or reverse/-) the read was aligned to 
```{r}
strand(bamDatFiltered)[1:10]
```

Load genomic data
```{r}
library(BSgenome.Dmelanogaster.UCSC.dm3)
```

To get the lengths of the chromosomes in the genome
```{r}
(chrSizes <- seqlengths(Dmelanogaster))
```

Load library and specifiy Drosophila database
```{r}
library(biomaRt)
ensembl <- useMart("ensembl", dataset = "dmelanogaster_gene_ensembl")
```


To define a set of fields we are interested for each gene
```{r}
annotation.fields <- c("ensembl_gene_id", "strand", "chromosome_name", "start_position", "end_position")
```

To restrict to chr2L
```{r}
gene.annotation <- getBM(annotation.fields, mart = ensembl, filters = "chromosome_name", values = c("2L"))
str(gene.annotation)
```

To check we only downloaded annotations from chr2L
```{r}
levels(as.factor(gene.annotation$chromosome))
```

Add chr to the annotation data
```{r}
gene.annotation$chromosome <- paste("chr", gene.annotation$chromosome_name, 
    sep = "")
levels(as.factor(gene.annotation$chromosome))
```

To store the gene annotation information in an IRanges object
```{r}
gene.range <- RangedData(IRanges(start = gene.annotation$start_position, end = gene.annotation$end_position), 
    space = gene.annotation$chromosome, strand = gene.annotation$strand, gene = gene.annotation$ensembl_gene_id, 
    universe = "Dm3")
show(gene.range)
```

To find out how many bases cover each position in every chromosome
```{r}
(cover <- coverage(bamDatFiltered, width = chrSizes))
```

To aggregate the coverage for each gene
```{r}
gene.coverage <- aggregate(cover[match(names(gene.range), names(cover))], ranges(gene.range), sum)
```

To find the number of reads covering each gene
```{r}
gene.coverage <- ceiling(gene.coverage/unique(width(bamDat)))
gene.coverage
```

To restrict to chr2L
```{r}
length(gene.coverage[["chr2L"]])
length(ranges(gene.range)$chr2L)
```

To build a count table and store it in a data frame
```{r}
countTable <- data.frame(chromosome = gene.range$space, gene_start = start(gene.range$ranges), 
    gene_end = end(gene.range$ranges), strand = gene.range$strand, gene = gene.range$gene, 
    count = as.vector(gene.coverage[["chr2L"]]))
dim(countTable)
head(countTable)
```

To use RPKM (reads per kilobase per million mapped reads)
```{r}
countTable <- data.frame(chromosome = gene.range$space, gene_start = start(gene.range$ranges), 
gene_end = end(gene.range$ranges), strand = gene.range$strand, gene = gene.range$gene, 
count = as.vector(gene.coverage[["chr2L"]]), RPKM = (as.vector(gene.coverage[["chr2L"]])/(end(gene.range$ranges) -                                                                                             start(gene.range$ranges))) * (1e+09/length(bamDat)))
head(countTable)
```

